### PR TITLE
drivers: video: ov7725: add sensor driver for ov7725

### PR DESCRIPTION
**Pull request by lgl88911**
This patch adds the driver for Omnivision OV7725
Color CMOS VGA Sensor.

The driver currently provides a simple capture
function, the output format only provides
RGB565,640x480.

Signed-off-by: Frank Li <lgl88911@163.com>

### CODE REVIEW

**Comment by loicpoulain** [comment]
It looks interesting, but could you please split it in several commits, at least one for the dts changes, one for the new ov driver and one for the sample.

**Comment by lgl88911** [comment]
> @lgl88911, It looks interesting, but could you please split it in several commits, at least one for the dts changes, one for the new ov driver and one for the sample.
Thanks for review, the PR will be close and I divided this commit into three: 
1. OV7725 driver https://github.com/zephyrproject-rtos/zephyr/pull/30787
In order to avoid CI fail, After https://github.com/zephyrproject-rtos/zephyr/pull/30787 be merged, the following two branches will make a pull request.
2. mm_swiftio support OV7725
3. sample

**Comment by loicpoulain** [comment]
> > @lgl88911, It looks interesting, but could you please split it in several commits, at least one for the dts changes, one for the new ov driver and one for the sample.
> > Thanks for review, the PR will be close and I divided this commit into three:
> 
>     1. OV7725 driver #30787
>        In order to avoid CI fail, After #30787 be merged, the following two branches will make a pull request.
> 
>     2. mm_swiftio support OV7725, https://github.com/lgl88911/zephyr/tree/mm_swiftio_ov7725
> 
>     3. sample，https://github.com/lgl88911/zephyr/tree/add_ov7725_sample

No, keep one branch and one PR/MR, don't need to close this one, I just wanted you split your changes in multiple git commits not in multiple branch.

**Comment by lgl88911** [comment]
@loicpoulain Thanks for your explain, I split it to three commit, Please help review again

File: dts/bindings/vendor-prefixes.txt [hunk 1]
```diff
@@ -306,6 +306,7 @@ okaya	Okaya Electric America, Inc.
 oki	Oki Electric Industry Co., Ltd.
 olimex	OLIMEX Ltd.
 olpc	One Laptop Per Child
+omnivision OmniVision Technologies Co., Ltd.
```
Comment (loicpoulain) [inline dts/bindings/vendor-prefixes.txt:309]: Under Linux OmniVision Technologies vendor prefix is 'ovti' , I would prefer that one for consistency.

File: dts/bindings/video/omnivision,ov7725.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# Copyright (c) 2020, FrankLi Limited
+# SPDX-License-Identifier: Apache-2.0
+
+description: OV7725 CMOS video sensor
+
+compatible: "omnivision,ov7725"
+
```
Comment (loicpoulain) [inline dts/bindings/video/omnivision,ov7725.yaml:7]: ovti,ov7725

File: drivers/video/ov7725.c [hunk 3]
```diff
@@ -0,0 +1,651 @@
...
+				gpio_name);
+		}
+	}
+
+	drv_data->i2c_addr = DT_INST_REG_ADDR(0);
+
+	return ov7725_init(dev);
+}
+
+DEVICE_AND_API_INIT(ov7725, DT_INST_LABEL(0), &ov7725_init_0,
```
Comment (loicpoulain) [inline drivers/video/ov7725.c:648]: That needs to be converted to the new DT device macro: DEVICE_AND_API_INIT -> DEVICE_DT_INST_DEFINE
cf https://github.com/zephyrproject-rtos/zephyr/pull/30779/commits/fc110b67ea66c77c43565a014583ab00012ff409

**Review by loicpoulain — CHANGES_REQUESTED** [review]
Overall, it looks good, just few comments to address.

File: dts/bindings/video/ovti,ov7725.yaml [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+# SPDX-License-Identifier: Apache-2.0
+
+description: OV7725 CMOS video sensor
+
+compatible: "ovti,ov7725"
+
+properties:
+    reset-gpios:
+      type: phandle-array
+      required: false
```
Comment (loicpoulain) [inline dts/bindings/video/ovti,ov7725.yaml:11]: Is it really optional? the driver seems to unconditionnaly request it.

File: dts/bindings/vendor-prefixes.txt (hunk 1 above)
Comment (lgl88911) [reply dts/bindings/vendor-prefixes.txt]: Very reasonable. Fixed

File: dts/bindings/video/omnivision,ov7725.yaml (hunk 2 above)
Comment (lgl88911) [reply dts/bindings/video/omnivision,ov7725.yaml]: Very reasonable , fFxed

File: drivers/video/ov7725.c (hunk 3 above)
Comment (lgl88911) [reply drivers/video/ov7725.c]: Change is so fast @@. converted done.

File: dts/bindings/video/ovti,ov7725.yaml (hunk 4 above)
Comment (lgl88911) [reply dts/bindings/video/ovti,ov7725.yaml:11]: Yes, If reset is connected to the reset circuit of the hardware, there is no need to use GPIO to control it.
If the gpio name cannot be found in the driver, gpio reset will not be performed
```
char *gpio_name = DT_INST_GPIO_LABEL(0, reset_gpios);

if (gpio_name) {
		drv_data->reset_gpio = device_get_binding(gpio_name);
		.....
}

...

if (drv_data->reset_gpio) {
		ret = gpio_pin_configure(drv_data->reset_gpio,
					 drv_data->reset_pin,
					 GPIO_OUTPUT_ACTIVE |
					 drv_data->reset_flags);
		if (ret) {
			return ret;
		}

		gpio_pin_set(drv_data->reset_gpio,
			     drv_data->reset_pin,
			     0);
		k_sleep(K_MSEC(1));
		gpio_pin_set(drv_data->reset_gpio,
			     drv_data->reset_pin,
			     1);
		k_sleep(K_MSEC(1));
	}
```

File: drivers/video/ov7725.c [hunk 5]
```diff
@@ -0,0 +1,651 @@
...
+#define OV7725_HUECOS     0xA9U
+#define OV7725_HUESIN     0xAAU
+#define OV7725_SIGN       0xABU
+#define OV7725_DSPAUTO    0xACU
+
+#define OV7725_COM10_VSYNC_NEG_MASK    (1U << 1U)
+#define OV7725_COM10_HREF_REVERSE_MASK (1U << 3U)
+#define OV7725_COM10_PCLK_REVERSE_MASK (1U << 4U)
+#define OV7725_COM10_PCLK_OUT_MASK     (1U << 5U)
+#define OV7725_COM10_DATA_NEG_MASK     (1U << 7U)
```
Comment (MaureenHelm) [inline drivers/video/ov7725.c:176]: Recommend using `BIT()` from `include/sys/util.h`

File: drivers/video/ov7725.c [hunk 6]
```diff
@@ -0,0 +1,651 @@
...
+	{ .input_clk = 13000000, .framerate = 14,
+		.clkrc = 0x01, .com4 = 0x41, .dm_lnl = 0x46 },
+};
+
+
+static const struct ov7725_pixel_format ov7725_pf_configs[] = {
+	{ .pixel_format = VIDEO_PIX_FMT_RGB565, .com7 = (1 << 2) | (2) }
+};
+
+static struct ov7725_reg ov7725_init_reg_tb[] = {
```
Comment (MaureenHelm) [inline drivers/video/ov7725.c:237]: `const`?

File: drivers/video/ov7725.c [hunk 7]
```diff
@@ -0,0 +1,651 @@
...
+
+	msgs[0].buf = (uint8_t *)&reg_addr;
+	msgs[0].len = 1;
+	msgs[0].flags = I2C_MSG_WRITE;
+
+	msgs[1].buf = (uint8_t *)&value;
+	msgs[1].len = 1;
+	msgs[1].flags = I2C_MSG_WRITE | I2C_MSG_STOP;
+
+	return i2c_transfer(drv_data->i2c, msgs, 2, drv_data->i2c_addr);
```
Comment (MaureenHelm) [inline drivers/video/ov7725.c:331]: Is there a reason you can't use `i2c_reg_write_byte()` instead?

File: drivers/video/ov7725.c [hunk 8]
```diff
@@ -0,0 +1,651 @@
...
+
+	msgs[0].buf = (uint8_t *)&reg_addr;
+	msgs[0].len = 1;
+	msgs[0].flags = I2C_MSG_WRITE | I2C_MSG_STOP;
+
+	msgs[1].buf = (uint8_t *)value;
+	msgs[1].len = 1;
+	msgs[1].flags = I2C_MSG_READ | I2C_MSG_STOP | I2C_MSG_RESTART;
+
+	return i2c_transfer(drv_data->i2c, msgs, 2, drv_data->i2c_addr);
```
Comment (MaureenHelm) [inline drivers/video/ov7725.c:348]: Replace with `i2c_reg_read_byte()`

File: drivers/video/ov7725.c [hunk 9]
```diff
@@ -0,0 +1,651 @@
...
+	int ret;
+	uint8_t set_value;
+
+	ret = ov7725_read_reg(dev, reg_addr, &set_value);
+
+	if (ret == 0) {
+		set_value = (set_value & (~clear_mask)) |
+				(set_value & clear_mask);
+		ret = ov7725_write_reg(dev, reg_addr, set_value);
+	}
```
Comment (MaureenHelm) [inline drivers/video/ov7725.c:365]: Replace with `i2c_reg_update_byte()`

File: boards/arm/mm_swiftio/pinmux.c [hunk 10]
```diff
@@ -20,30 +20,30 @@ static gpio_pin_config_t enet_gpio_config = {
...
+ * Pull Up / Down Config. Field: 47K Ohm Pull Up
+ * Hyst. Enable Field: Hysteresis Enabled.
  */
 
 static void mm_swiftio_usdhc_pinmux(
 	uint16_t nusdhc, bool init,
 	uint32_t speed, uint32_t strength)
 {
 	uint32_t cmd_data = IOMUXC_SW_PAD_CTL_PAD_SPEED(speed) |
-				IOMUXC_SW_PAD_CTL_PAD_SRE_MASK |
```
Comment (MaureenHelm) [inline boards/arm/mm_swiftio/pinmux.c:36]: Please move the unrelated whitespace changes to a separate commit

File: samples/video/capture/boards/mm_swiftio.conf [hunk 11]
```diff
@@ -0,0 +1,2 @@
+CONFIG_VIDEO_OV7725=y
+CONFIG_I2C=y
```
Comment (MaureenHelm) [inline samples/video/capture/boards/mm_swiftio.conf:2]: Please add this board to `samples/video/capture/sample.yaml` to make sure we build it in CI

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Thanks @lgl88911 !

File: drivers/video/ov7725.c (hunk 5 above)
Comment (lgl88911) [reply drivers/video/ov7725.c]: Thanks for review, Fixed

File: drivers/video/ov7725.c (hunk 6 above)
Comment (lgl88911) [reply drivers/video/ov7725.c]: Yes, lost const, Fixed

File: drivers/video/ov7725.c (hunk 7 above)
Comment (lgl88911) [reply drivers/video/ov7725.c:331]: Although i2c_reg_write_byte can achieve the same purpose, I hope it is the similar as ov7725_read_reg.

File: drivers/video/ov7725.c (hunk 8 above)
Comment (lgl88911) [reply drivers/video/ov7725.c:348]: OV7725 uses SCCB to communicate. SCCB is similar to I2C, but not exactly the same. When SCCB reads the register, an I2C stop bit is required after the register address of OV7725 is written, and then the read operation can be performed. i2c_reg_read_byte call i2c_write_read, Its implementation lacks a stop bit
```
{
	struct i2c_msg msg[2];

	msg[0].buf = (uint8_t *)write_buf;
	msg[0].len = num_write;
	msg[0].flags = I2C_MSG_WRITE;

	msg[1].buf = (uint8_t *)read_buf;
	msg[1].len = num_read;
	msg[1].flags = I2C_MSG_RESTART | I2C_MSG_READ | I2C_MSG_STOP;

	return i2c_transfer(dev, msg, 2, addr);
}
```

File: drivers/video/ov7725.c (hunk 9 above)
Comment (lgl88911) [reply drivers/video/ov7725.c:365]: The reason is similar to ov7725_read_reg, i2c_reg_update_byte will fail when reading SCCB.

File: boards/arm/mm_swiftio/pinmux.c (hunk 10 above)
Comment (lgl88911) [reply boards/arm/mm_swiftio/pinmux.c:36]: Thanks for review, Fixed

File: samples/video/capture/boards/mm_swiftio.conf (hunk 11 above)
Comment (lgl88911) [reply samples/video/capture/boards/mm_swiftio.conf:2]: I didn’t find any documentation about sample.yaml. I added one by referring to the existing example. It doesn’t seem to cause CI check fail, but I’m not sure that CI checked mm_swiftio or not. Please help to review whether this modification is correct. Thanks so much!

File: samples/video/capture/sample.yaml [hunk 12]
```diff
@@ -6,3 +6,8 @@ tests:
     tags: video
     platform_allow: mimxrt1064_evk
```
Comment (loicpoulain) [inline samples/video/capture/sample.yaml:7]: That is a list of allowed boards so you can directly append mm_swiftio here.

**Review by loicpoulain — CHANGES_REQUESTED** [review]

File: drivers/video/ov7725.c (hunk 8 above)
Comment (loicpoulain) [reply drivers/video/ov7725.c:348]: Would be nice to add the SCCB explanation as an inline comment.

File: samples/video/capture/sample.yaml (hunk 12 above)
Comment (lgl88911) [reply samples/video/capture/sample.yaml]: Thanks so much, fixed

File: drivers/video/ov7725.c (hunk 8 above)
Comment (lgl88911) [reply drivers/video/ov7725.c:352]: Thanks for suggest, Added it.

**Review by loicpoulain — APPROVED** [review]

**Comment by lgl88911** [comment]
@MaureenHelm  Can you help review again,Thanks

**Review by MaureenHelm — APPROVED** [review]
